### ISSUE TITLE
suggest adding #[test] on otherwise-dead code that looks like a test?

### ISSUE THREAD

**Issue by sourcefrog** [rust-lang/rust#33166]
On this code:

``` rust
#[cfg(test)]
mod tests {

    pub fn test_addition() {
        assert_eq!(2+2, 4);
    }
}
```

I currently get 

```
function is never used: `test_addition`, #[warn(dead_code)] on by default
```

What I need here is a `#[test]` before the function of course.  Perhaps rustc could suggest this if some of the following hints are seen
- it's in `#[cfg(test)]`
- it's in a submodule `tests`
- the function's called `test_*`

I don't know if making suggestions based on names is considered unhygienic, but this would remove a small speed bump for new users.

**Comment by GuillaumeGomez** [comment]
That's maybe a bit too much? The idea is good but I'm afraid it might overload the output. Let's see what others think about it.

**Comment by steveklabnik** [comment]
> the function's called test_* 

In general, naming your function with `test` doesn't seem idiomatic to me; it leads to duplicated output with the test runner.

**Comment by Mark-Simulacrum** [comment]
I personally think this is better suited as e.g. a clippy lint and doesn't really make sense for a compiler lint as there'd be too many false positives. It's also true that having helper functions inside `cfg(test)` is pretty common (at least in my experience), so that's another source of false positives. I'm going to go ahead and close this in favor of adding this to clippy and then if there's wide usage it can eventually be migrated into rustc through the RFC process.

**Comment by sourcefrog** [comment]
Sounds good
